File: archive/speedgrader/app/src/main/java/com/instructure/speedgrader/dialogs/TimePickerFragment.java
/*
 * Copyright (C) 2016 - present  Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.speedgrader.dialogs;

import android.app.Dialog;
import android.app.TimePickerDialog;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.text.format.DateFormat;
import android.widget.TimePicker;

import com.instructure.speedgrader.views.DateTextView;

import java.util.Calendar;
import java.util.Date;

public class TimePickerFragment extends DialogFragment implements TimePickerDialog.OnTimeSetListener{

    private Calendar calendar;

    public static TimePickerFragment getInstance(DateTextView dateTextView){

        TimePickerFragment datePickerFragment = new TimePickerFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable("dtv", dateTextView);
        datePickerFragment.setArguments(bundle);

        return datePickerFragment;
    }

    private DateTextView dateTextView;

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        // Use the current date as the default date in the picker
        dateTextView = (DateTextView) getArguments().getSerializable("dtv");
        Date date = dateTextView.getDate();

        //Error Checking
        if(date == null){
            date = new Date();
        }

        //DatePickerDialog expects data in a way that calendar provides natively.
        calendar = Calendar.getInstance();
        calendar.setTime(date);

        boolean is24HourDateFormat = DateFormat.is24HourFormat(getActivity());

        return new TimePickerDialog(getActivity(), this, calendar.get(Calendar.HOUR_OF_DAY),calendar.get(Calendar.MINUTE), is24HourDateFormat);
    }


    //We need to prevent duplicates. If the user leaves the page, we don't want two DatePickerFragments.
    @Override
    public void onPause(){
        super.onPause();
        dismiss();
    }

    @Override
    public void onTimeSet(TimePicker timePicker, int hourOfDay, int minute) {
        calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
        calendar.set(Calendar.MINUTE, minute);

        dateTextView.setDate(calendar.getTime());
    }
}

File: apps/student/src/main/java/com/instructure/student/fragment/TimePickerFragment.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.instructure.student.fragment

import android.app.Dialog
import android.app.TimePickerDialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.format.DateFormat
import android.widget.TimePicker
import androidx.fragment.app.DialogFragment
import java.util.*

class TimePickerFragment : DialogFragment(), TimePickerDialog.OnTimeSetListener {

    var timePickerListener: TimePickerFragmentListener? = null

    var cancelListener: TimePickerCancelListener? = null

    private var hour = -1
    private var minute = -1


    public interface TimePickerFragmentListener {
        fun onTimeSet(hourOfDay: Int, minute: Int)
    }

    public interface TimePickerCancelListener {
        fun onCancel()
    }

    fun notifyTimePickerCancelListener() {
        if (cancelListener != null) {
            cancelListener!!.onCancel()
        }
    }

    fun notifyTimePickerListener(hourOfDay: Int, minute: Int) {
        if (this.timePickerListener != null) {
            this.timePickerListener!!.onTimeSet(hourOfDay, minute)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        // Use the current time as the default values for the picker
        val c = Calendar.getInstance()
        if (hour != -1 && minute != -1) {
            return TimePickerDialog(requireActivity(), this, hour, minute,
                    DateFormat.is24HourFormat(activity))
        } else {
            return TimePickerDialog(requireActivity(), this, c.get(Calendar.HOUR_OF_DAY), c.get(Calendar.MINUTE),
                    DateFormat.is24HourFormat(activity))
        }
    }

    override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
        notifyTimePickerListener(hourOfDay, minute)
    }

    override fun onCancel(dialog: DialogInterface?) {
        super.onCancel(dialog)
        notifyTimePickerCancelListener()
    }

    companion object {

        fun newInstance(listener: TimePickerFragmentListener, cancelListener: TimePickerCancelListener): TimePickerFragment {
            val fragment = TimePickerFragment()
            fragment.timePickerListener = listener
            fragment.cancelListener = cancelListener
            return fragment
        }

        fun newInstance(listener: TimePickerFragmentListener, hour: Int, minute: Int, cancelListener: TimePickerCancelListener): TimePickerFragment {
            val fragment = TimePickerFragment()
            fragment.timePickerListener = listener
            fragment.cancelListener = cancelListener
            fragment.hour = hour
            fragment.minute = minute
            return fragment
        }
    }
}
